configfile: "config/config.yml"

myinput=list()
myoutput=list()


if "exome" in config["tech"] and "rna" in config["tech"]:
    myinput.append(expand("results_{sample_id}/{tec}/alignment/{tec}.positionsort.bam",tec="exome",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/filtration/snps_het.vcf.gz",tec="exome",sample_id=config["sample_name"]))
    myinput.append(expand("results_{sample_id}/{tec}/alignment/{tec}.splitted.bam",tec="rna",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/filtration/snps_het.vcf.gz",tec="rna",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec="rna",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/rna/transcripts_quant/quant.sf",sample_id=config["sample_name"]))
elif "exome" in config["tech"]:
    myinput.append(expand("results_{sample_id}/{tec}/alignment/{tec}.positionsort.bam",tec="exome",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/filtration/snps_het.vcf.gz",tec="exome",sample_id=config["sample_name"]))
elif "rna" in config["tech"]:
    myinput.append(expand("results_{sample_id}/{tec}/alignment/{tec}.splitted.bam",tec="rna",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/filtration/snps_het.vcf.gz",tec="rna",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec="rna",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/rna/transcripts_quant/quant.sf",sample_id=config["sample_name"]))

if "atac" in config["tech"]:
    myoutput.append(expand("results_{sample_id}/atac/peaks/annotated.tsv",sample_id=config["sample_name"]))
    myoutput.append(expand("results_{sample_id}/{tec}/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec="atac",sample_id=config["sample_name"]))

if "seesaw" in config["extra_analysis"]:
    myoutput.append(expand("results_{sample_id}/seesaw/salmon/quant.sf",sample_id=config["sample_name"]))

myoutput.append(expand("results_{sample_id}/phased/shapeit_whatshap.vcf.gz",sample_id=config["sample_name"]))
myoutput.append(expand("results_{sample_id}/merged_vcf/snps_het.vcf.gz",sample_id=config["sample_name"]))
myoutput.append(expand("results_{sample_id}/phased/manual_refinment.vcf.gz",sample_id=config["sample_name"]))

rule all:
    input:
        myoutput
        #"results/exome/recalibration/exome.positionsort.gatkgroup.bam",
        #expand("results/{tec}/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec=config["tech"])
        #expand("results/exome/ASE{chrom}",chrom=config["chromosomes_to_phase"],tec=config["tech"])

include: "rules/alignment.smk"
include: "rules/qc_aligned.smk"
include: "rules/hap_caller.smk"
include: "rules/phasing.smk"
include: "rules/ASE.smk"
include: "rules/peak_calling.smk"
include: "rules/add_genome.smk"
include: "rules/seesaw.smk"