# ASTRA bulk pipeline - currently testing

Snakemake workflow for the analysis of allele specific expression and/or chromatin accessibility from sequencing data.

<img src="./pipeline.svg">

## Input files needed

As input are needed the gzipped fastq file of *one* sample.
It is possible to use either RNA-seq and/or ATAC-seq and eventual WES fastq files and/or additional known SNPs (vcf file format).

## Reference files preparation

### Genome gtf and fasta files
Gtf and fasta reference files will be used several times thoughout the pipeline, for the alignment, variant calling, feature counting etc.

Files available on [Ensembl web site](https://www.ensembl.org/Homo_sapiens/Info/Index).
GRCh38 download example:

```bash 
wget http://ftp.ensembl.org/pub/release-107/gtf/homo_sapiens/Homo_sapiens.GRCh38.107.chr.gtf.gz
gzip -d Homo_sapiens.GRCh38.107.chr.gtf.gz
wget http://ftp.ensembl.org/pub/release-107/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
gzip -d Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
```
Add file paths to [config file](config/config.yml) in `genome_gtf` and `genome_fa`

### ðŸ§¬ Hisat2
The alignment is performed using Hisat2. Hisat2 specific index file should be created.

Prepare [hisat2](https://www.nature.com/articles/s41587-019-0201-4) index files available [here](http://daehwankimlab.github.io/hisat2/download/) for download or preparation instructions with custom reference available [here](http://daehwankimlab.github.io/hisat2/howto/#build-hgfm-index-with-snps-and-transcripts). 

For an example look at this [file](hisat_indexes.sh).

Add file paths to [config file](config/config.yml) in `hisat_index`

### ðŸ§¬ Haptree-X
Haplotype phasing is performed using 3 different tools, WhatsHap, Shapeit and Haptree-X. While the other methods are automaticaly included in the pipeline through conda environments or docker images, Haptree-X should be installed locally. In particular, you would just need to download the last released executable file, an example follows.

The Haptree-X excecutable file can be downloaded from [here](https://github.com/0xTCG/haptreex/releases).

Add path to [config file](config/config.yml) in `haptreex_exe`

### ðŸ§¬ Lorals Tool for Allele-Specific Expression (ASE)

The **Lorals tool** is utilized here for **Allele-Specific Expression (ASE)** analysis specifically on **long-reads RNA-seq** data.

---

#### âš ï¸ Important Note: Modified Tool Version

The original Lorals tool's maintenance has been discontinued by its developers. To ensure successful operation, the tool required **slight modifications** to its source code.

> **The functioning, modified version is currently located in the `resources` directory.**

### Installation

To install this version of the tool, you must unzip the lorals.tar.gz file and compile the included `setup.py` file. Please follow the standard procedure, which is also outlined in the [original Lorals GitHub documentation](https://github.com/LappalainenLab/lorals?tab=readme-ov-file#lorals):

```bash
tar -xzvf lorals.tar.gz
cd lorals
python3 setup.py install
```

Lorals original page can be found [here](https://github.com/LappalainenLab/lorals?tab=readme-ov-file#lorals).

### ðŸ§¬ Other reference files 

Reference file of phased variants for variant calling with Shapeit4 available [here](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working).

**Be careful to have the same chromosome notation in all the reference and target files (chr1 or 1).**

<details><summary>GRCh38 download example </summary>
<p> 


```bash 
for i in {1..22} X; do wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20220422_3202_phased_SNV_INDEL_SV/1kGP_high_coverage_Illumina.chr$i.filtered.SNV_INDEL_SV_phased_panel.vcf.gz; done

for i in {1..22} X; do wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20220422_3202_phased_SNV_INDEL_SV/1kGP_high_coverage_Illumina.chr$i.filtered.SNV_INDEL_SV_phased_panel.vcf.gz.tbi; done
```

</p>
</details>


Moreover, for the RNA/WES Hard Filtering step is required a dbSNP.
<details><summary>GRCh38 download example </summary>
<p> 


```bash 
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz.tbi
```

</p>
</details>

## How to run

```bash
snakemake --cores [cores_number] --use-conda
```

Once the config file is fully completed you can run the entire pipeline by just typing the command above. 

## Results architecture
<details><summary> Output architecture </summary>
<p> 
  
```bash
results/
â”œâ”€â”€ exome
â”‚Â Â  â”œâ”€â”€ filtration
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ filtered.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ filtered.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels_filtered.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels_filtered.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ metrics.variant_calling_detail_metrics
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ metrics.variant_calling_summary_metrics
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ sample.txt
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps_filtered.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps_filtered.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps_het.vcf.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ snps_het.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ haplotypeCaller
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ exome.g.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ exome.vcf.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ exome.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ prephasing
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ pre_phased.vcf.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ pre_phased.vcf.gz.tbi
â”‚Â Â  â””â”€â”€ recalibration
â”‚Â Â      â”œâ”€â”€ exome.recal.bai
â”‚Â Â      â”œâ”€â”€ exome.recal.bam
â”‚Â Â      â””â”€â”€ recal_data.table
â”œâ”€â”€ rna
â”‚Â Â  â”œâ”€â”€ ASEX
â”‚Â Â  â”‚Â Â  â””â”€â”€ rna.table
â”‚Â Â  â”œâ”€â”€ alignment
â”‚Â Â  â”‚Â Â  â””â”€â”€ rna.splitted.bai
â”‚Â Â  â”œâ”€â”€ filtration
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ filtered.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ filtered.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels_filtered.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ indels_filtered.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ metrics.variant_calling_detail_metrics
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ metrics.variant_calling_summary_metrics
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ sample.txt
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps_filtered.vcf.gz
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps_filtered.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ snps_het.vcf.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ snps_het.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ haplotypeCaller
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ rna.g.vcf.gz.tbi
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ rna.vcf.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ rna.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ prephasing
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ pre_phased.vcf.gz
â”‚Â Â  â”‚Â Â  â””â”€â”€ pre_phased.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ recalibration
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ recal_data.table
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ rna.recal.bai
â”‚Â Â  â”‚Â Â  â””â”€â”€ rna.recal.bam
â”‚Â Â  â””â”€â”€ transcripts_quant
â”‚Â Â      â”œâ”€â”€ logs
â”‚Â Â      â”‚Â Â  â””â”€â”€ salmon_quant.log
â”‚Â Â      â””â”€â”€ quant.sf
â”œâ”€â”€ merged_vcf
â”‚Â Â  â”œâ”€â”€ snps_het.vcf.gz
â”‚Â Â  â””â”€â”€ snps_het.vcf.gz.tbi
â”œâ”€â”€ phased
â”‚Â Â  â”œâ”€â”€ haptreex.tsv
â”‚Â Â  â”œâ”€â”€ manual_refinment.vcf.gz
â”‚Â Â  â”œâ”€â”€ manual_refinment.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ pre_phased.vcf.gz
â”‚Â Â  â”œâ”€â”€ pre_phased.vcf.gz.tbi
â”‚Â Â  â”œâ”€â”€ shapeit_whatshap.vcf.gz
â”‚Â Â  â””â”€â”€ shapeit_whatshap.vcf.gz.tbi
â””â”€â”€ seesaw
    â”œâ”€â”€ salmon
    â”‚Â Â  â”œâ”€â”€ logs
    â”‚Â Â  â”‚Â Â  â””â”€â”€ salmon_quant.log
    â”‚Â Â  â””â”€â”€ quant.sf
    â””â”€â”€ transcripts.fa
```

</p>
</details>
